### PR TITLE
Adapt TransactionFailureReason to iota-core

### LINKED ISSUE #1945: Replicate iota-core FailureReasons
https://github.com/iotaledger/iota.go/pull/672

### PR DESCRIPTION
**Pull request by thibault-martinez**
Fixes https://github.com/iotaledger/iota-sdk/issues/1945

### CODE REVIEW

File: sdk/src/types/block/semantic/error.rs [hunk 1]
```diff
@@ -144,32 +237,71 @@ impl TryFrom<u8> for TransactionFailureReason {
...
-            18 => Self::MissingStakingFeature,
-            19 => Self::FailedToClaimStakingReward,
-            20 => Self::FailedToClaimDelegationReward,
-            21 => Self::TransactionCapabilityNativeTokenBurningNotAllowed,
-            22 => Self::TransactionCapabilityManaBurningNotAllowed,
-            23 => Self::TransactionCapabilityAccountDestructionNotAllowed,
-            24 => Self::TransactionCapabilityAnchorDestructionNotAllowed,
-            25 => Self::TransactionCapabilityFoundryDestructionNotAllowed,
-            26 => Self::TransactionCapabilityNftDestructionNotAllowed,
+            0 => Self::None,
```
Comment (Thoralf-M) [inline sdk/src/types/block/semantic/error.rs]: Can't we just use https://docs.rs/strum/latest/strum/derive.FromRepr.html?

Comment (thibault-martinez) [reply sdk/src/types/block/semantic/error.rs]: I'll try

Comment (thibault-martinez) [reply sdk/src/types/block/semantic/error.rs]: Done

File: sdk/src/types/block/semantic/unlock.rs [hunk 2]
```diff
@@ -24,44 +24,49 @@ impl SemanticValidationContext<'_> {
...
+                        return Err(TransactionFailureReason::SemanticValidationFailed);
                     }
                     if !self.unlocked_addresses.contains(address) {
-                        return Err(TransactionFailureReason::InvalidInputUnlock);
+                        // TODO
+                        return Err(TransactionFailureReason::SemanticValidationFailed);
                     }
                 } else {
-                    return Err(TransactionFailureReason::InvalidInputUnlock);
+                    // TODO
```
Comment (DaughterOfMars) [inline sdk/src/types/block/semantic/unlock.rs]: issue?

File: sdk/src/types/block/semantic/error.rs (hunk 1 above)
Comment (Thoralf-M) [reply sdk/src/types/block/semantic/error.rs]: Nice

File: sdk/src/types/block/semantic/unlock.rs (hunk 2 above)
Comment (thibault-martinez) [reply sdk/src/types/block/semantic/unlock.rs]: Added

**Review by DaughterOfMars — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Pull request merged**